### DISCUSSION TITLE
ExProto: ability to specify endpoint via domain name.

### DISCUSSION THREAD

**Discussion by paraskun** [emqx/emqx#14361] [Q&A] [answered]
At this point, we can configure the gateway to talk to our gRPC service by specifying an IP address, for example:

`http://10.5.0.5:9111`

Working with Docker in general and Swarm in particular, we are faced with dynamic addressing in our networks, so it will be very convenient to specify a domain name (container name) instead of an IP address.

`http://container:9111`

Is there a way to achieve this?

**Comment by id** [comment]
@paraskun are you saying that using hostname for ExHook URL does not work?

**Reply by paraskun** [reply]
If we talk about ExProto and ExHook as interchangeable concepts, then yes.
Here is the error message sent by EMQX when trying to specify an endpoint containing a domain name:

`
emqx-1  | 2024-12-09T08:09:39.354385+00:00 [error] msg: load_gateway_crashed, reason: {error,badarg}, stacktrace: [{io_lib,format,["~s://~s:~w",[http,{error,einval},9111]],[{file,"io_lib.erl"},{line,189}]},{emqx_gateway_exproto,compose_http_uri,3,[{file,"emqx_gateway_exproto.erl"},{line,231}]},{emqx_gateway_exproto,start_grpc_client_channel,2,[{file,"emqx_gateway_exproto.erl"},{line,206}]},{emqx_gateway_exproto,on_gateway_load,2,[{file,"emqx_gateway_exproto.erl"},{line,61}]},{emqx_gateway_insta_sup,cb_gateway_load,1,[{file,"emqx_gateway_insta_sup.erl"},{line,453}]},{emqx_gateway_insta_sup,init,1,[{file,"emqx_gateway_insta_sup.erl"},{line,118}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,980}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,935}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}], gateway_name: exproto, gateway: #{name => exproto,status => stopped,config => #{handler => #{address => <<"http://extd:9111">>,ssl_options => #{depth => 10,hibernate_after => 5000,verify => verify_none,enable => false,ciphers => [],versions => ['tlsv1.3','tlsv1.2'],log_level => notice,secure_renegotiate => true,reuse_sessions => true},service_name => 'ConnectionUnaryHandler'},server => #{bind => {{0,0,0,0},9100},ssl_options => #{depth => 10,hibernate_after => 5000,verify => verify_none,ciphers => [],versions => ['tlsv1.3','tlsv1.2'],log_level => notice,client_renegotiation => true,fail_if_no_peer_cert => false,certfile => <<"${EMQX_ETC_DIR}/certs/cert.pem">>,keyfile => <<"${EMQX_ETC_DIR}/certs/key.pem">>,cacertfile => <<"${EMQX_ETC_DIR}/certs/cacert.pem">>,honor_cipher_order => true,secure_renegotiate => true,reuse_sessions => true,user_lookup_fun => {fun emqx_tls_psk:lookup/3,undefined},handshake_timeout => 15000}},enable => true,listeners => #{tcp => #{default => #{bind => 20041,tcp_options => #{buffer => 4096,high_watermark => 1048576,keepalive => "none",nodelay => false,reuseaddr => true,send_timeout => 15000,send_timeout_close => true,backlog => 1024,active_n => 100,nolinger => false},enable => true,acceptors => 16,proxy_protocol_timeout => 3000,proxy_protocol => false,max_connections => 1024000,access_rules => [],max_conn_rate => 1000,enable_authn => true}}},idle_timeout => 30000,mountpoint => <<>>,enable_stats => true},created_at => 1733731779350}
emqx-1  | 2024-12-09T08:09:39.355597+00:00 [error] crasher: initial call: emqx_gateway_insta_sup:init/1, pid: <0.3671.0>, registered_name: [], exit: {badarg,[{gen_server,init_it,6,[{file,"gen_server.erl"},{line,961}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}, ancestors: [exproto,emqx_gateway_sup,<0.3345.0>], message_queue_len: 0, messages: [], links: [<0.3651.0>], dictionary: [], trap_exit: true, status: running, heap_size: 10958, stack_size: 28, reductions: 20219; neighbours:
emqx-1  | 2024-12-09T08:09:39.356095+00:00 [warning] msg: cluster_rpc_apply_result, result: {error,{post_config_update,emqx_gateway_conf,badarg}}, kind: initiate, tnx_id: 1, entrypoint: <<"emqx:update_config/3">>
`

In this mess the address is `http://extd:9111`. When specifying the IP everything works fine.

PS: Sorry for the unstructured logs.

**Reply by id** [reply, marked as answer]
Got it. Yes, this looks like a bug. Thanks for the report, I'll create an internal issue to follow up.

**Reply by paraskun** [reply]
Thanks for your reply!
